cupy/core/core.pyx

```diff
@@ -3895,139 +3890,127 @@ cpdef ndarray matmul(ndarray a, ndarray b, ndarray out=None):
+    for i in range(0, ndim - 2):
+        a_sh = a._shape[i]
+        b_sh = b._shape[i]
+        if a_sh != b_sh and a_sh != 1 and b_sh != 1:
+            raise ValueError(
+                'operands could not be broadcast together with '
+                'remapped shapes')
 
-    view = a.view()
-    # TODO(niboshi): Confirm update_x_contiguity flags
-    view._set_shape_and_strides(a_shape, a_strides, True, True)
-    a = view
-    view = b.view()
-    # TODO(niboshi): Confirm update_x_contiguity flags
-    view._set_shape_and_strides(b_shape, b_strides, True, True)
-    b = view
+        if a_sh == 0 or b_sh == 0:
+            c_sh = 0
+        else:
+            c_sh = max(a_sh, b_sh)
+        batchCount *= c_sh
+        out_shape.append(c_sh)
+        if a_sh == 1 and c_sh > 1:
+            a._strides[i] = 0
+            a._shape[i] = c_sh
+            a._c_contiguous = a._f_contiguous = False
+            use_broadcast = True
 
-    out = ndarray(out_shape, dtype=dtype)
-    out.data.memset_async(0, out.nbytes)
-
-    out_view = out.view()
-    out_view_shape = out.shape
-    out_view_strides = out.strides
-    if a_part_outshape == ():
-        out_view_shape += (1,)
-        out_view_strides += (0,)
-    if b_part_outshape == ():
-        out_view_shape += (1,)
-        out_view_strides += (0,)
+        if b_sh == 1 and c_sh > 1:
+            b._strides[i] = 0
+            b._shape[i] = c_sh
+            b._c_contiguous = b._f_contiguous = False
+            use_broadcast = True
 
-    # TODO(niboshi): Confirm update_x_contiguity flags
-    out_view._set_shape_and_strides(
-        out_view_shape, out_view_strides, True, True)
+    if orig_a_ndim != 1:
+        out_shape.append(a_part_outshape)
+    if orig_b_ndim != 1:
+        out_shape.append(b_part_outshape)
 
     # (A B)^T = B^T A^T
     a, b = b, a
 
-    lda = a.shape[-1]
-    ldb = b.shape[-1]
-
-    *la, ka, n = a.shape
-    *lb, m, kb = b.shape
+    ka = a._shape[ndim - 2]
+    lda = n = a._shape[ndim - 1]
+    m = b._shape[ndim - 2]
+    ldb = kb = b._shape[ndim - 1]
 
     if ka != kb:
         raise ValueError(
             'shapes ({}) and ({}) not aligned'.format(
-                ','.join([str(_) for _ in orig_a_shape]),
-                ','.join([str(_) for _ in orig_b_shape])))
-    for la_, lb_ in zip(la, lb):
-        if not (la_ == lb_ or la_ == 1 or lb_ == 1):
-            raise ValueError(
-                'operands could not be broadcast together with '
-                'remapped shapes')
+                ','.join([str(_) for _ in orig_a.shape]),
+                ','.join([str(_) for _ in orig_b.shape])))
 
     if a.size == 0 or b.size == 0:
         return cupy.zeros(out_shape, ret_dtype)
 
-    batchCount = 1  # batchCount = numpy.prod(la)
-    for i in la:
-        batchCount *= i
+    out = ndarray(out_shape, dtype=dtype)
+
+    if orig_a_ndim == 1 or orig_b_ndim == 1:
+        out_view = out.view()
+        if orig_a_ndim == 1:
```

Comment:
At most one of `orig_a_ndim` and `orig_b_ndim` is 1 because of the early-return for `ndim <= 2`.